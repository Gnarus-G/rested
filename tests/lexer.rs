use insta::assert_debug_snapshot;

macro_rules! assert_lexes {
    ($input:literal) => {
        let lexer = rested::lexer::Lexer::new($input);

        insta::with_settings!({
             description => $input
        }, {
            assert_debug_snapshot!(lexer.into_iter().collect::<Vec<rested::lexer::Token>>());
        })
    }
}

#[test]
fn lex_put_patch_delete() {
    assert_lexes!(
        r#"
put /api {}
patch /api {}
delete /api {}
"#
    );
}

#[test]
fn lex_string_literals() {
    assert_lexes!(r#""hello""#);

    assert_lexes!(r#""hello"#);

    assert_lexes!(
        r#"
"hello
"world
"#
    );

    assert_lexes!(r#" "" "" ``"#);

    assert_lexes!(r#" { "Bearer token" } "#);

    assert_lexes!(
        r#"`
{
    stuff
}`

`
stuff"#
    );
}

#[test]
fn lex_bools() {
    assert_lexes!("true false");
}

#[test]
fn lex_numbers() {
    assert_lexes!("123124 1.0");
}

#[test]
fn lex_get_url() {
    assert_lexes!("get http://localhost");
}

#[test]
fn lex_get_url_with_header() {
    assert_lexes!("get http://localhost { header \"Authorization\" \"Bearer token\" }");
}

#[test]
fn lex_get_url_over_many_lines() {
    assert_lexes!("get\nhttp://localhost");

    assert_lexes!(
        r#"get 
    http://localhost 
{
}"#
    );
}

#[test]
fn lex_get_url_with_header_and_body() {
    assert_lexes!(
        r#"
post http://localhost { 
    header "Authorization" "Bearer token" 
    body "{neet: 1337}" 
}"#
    );
}

#[test]
fn lex_call_expression() {
    assert_lexes!(r#"env() env("stuff")"#);
}

#[test]
fn lex_template_literals() {
    assert_lexes!(r#"`stuff${"interpolated"}(things${env("dead_night")}` `dohickeys`"#);

    assert_lexes!(r#"`a${"temp"}` }}"#);

    assert_lexes!(
        r#"
let b = `asdf ${}`
"#
    );

    assert_lexes!(r#"` ${} ${} `"#);
    assert_lexes!(r#"`head ${} ${} tail`"#);

    assert_lexes!(
        r#"
let b = `asdf ${test} ${} ${"word"}`
"#
    );

    assert_lexes!(
        r#"
let b = `asdf ${}` 
let c = {}
"#
    );

    assert_lexes!(r#"`${test} ${df} `"#);
    assert_lexes!(r#"`asdf ${test} ${df} `"#);
}

#[test]
fn lex_json_object() {
    assert_lexes!(
        r#"
let obj = {
    key: "value",
    otherKey: ["val", "vlue"]
}
"#
    );
}
