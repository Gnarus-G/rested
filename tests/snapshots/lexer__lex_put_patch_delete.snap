---
source: tests/lexer.rs
description: "\nput /api {}\npatch /api {}\ndelete /api {}\n"
expression: "lexer.into_iter().collect::<Vec<rested::lexer::Token>>()"
---
[
    Put("put") at Position { value: 1, line: 1, col: 0 },
    Pathname("/api") at Position { value: 5, line: 1, col: 4 },
    LBracket("{") at Position { value: 10, line: 1, col: 9 },
    RBracket("}") at Position { value: 11, line: 1, col: 10 },
    Patch("patch") at Position { value: 13, line: 2, col: 0 },
    Pathname("/api") at Position { value: 19, line: 2, col: 6 },
    LBracket("{") at Position { value: 24, line: 2, col: 11 },
    RBracket("}") at Position { value: 25, line: 2, col: 12 },
    Delete("delete") at Position { value: 27, line: 3, col: 0 },
    Pathname("/api") at Position { value: 34, line: 3, col: 7 },
    LBracket("{") at Position { value: 39, line: 3, col: 12 },
    RBracket("}") at Position { value: 40, line: 3, col: 13 },
]
